test/Microsoft.AspNetCore.Sockets.Common.Tests/Internal/Formatters/TextMessageParserTests.cs

```diff
@@ -0,0 +1,115 @@
+            var reader = new BytesReader(buffer);
+
+            Assert.True(parser.TryParseMessage(ref reader, MessageFormat.Text, out var message));
+            Assert.Equal(reader.Index, buffer.Length);
+
+            MessageTestUtils.AssertMessage(message, messageType, payload);
+        }
+
+        [Theory]
+        [InlineData("0:B:;", new byte[0])]
+        [InlineData("8:B:q83vEg==;", new byte[] { 0xAB, 0xCD, 0xEF, 0x12 })]
+        [InlineData("8:B:q83vEjQ=;", new byte[] { 0xAB, 0xCD, 0xEF, 0x12, 0x34 })]
+        [InlineData("8:B:q83vEjRW;", new byte[] { 0xAB, 0xCD, 0xEF, 0x12, 0x34, 0x56 })]
+        public void ReadBinaryMessage(string encoded, byte[] payload)
+        {
+            var parser = new MessageParser();
+            var buffer = Encoding.UTF8.GetBytes(encoded);
+            var reader = new BytesReader(buffer);
+
+            Assert.True(parser.TryParseMessage(ref reader, MessageFormat.Text, out var message));
+            Assert.Equal(reader.Index, buffer.Length);
+
+            MessageTestUtils.AssertMessage(message, MessageType.Binary, payload);
+        }
+
+        [Theory]
+        [InlineData(0)] // Not chunked
+        [InlineData(4)]
+        [InlineData(8)]
+        public void ReadMultipleMessages(int chunkSize)
+        {
+            const string encoded = "0:B:;14:T:Hello,\r\nWorld!;1:C:A;12:E:Server Error;";
+            var parser = new MessageParser();
+            var data = Encoding.UTF8.GetBytes(encoded);
+            var buffer = chunkSize > 0 ?
+                data.ToChunkedReadOnlyBytes(chunkSize) :
+                new ReadOnlyBytes(data);
+
+            var reader = new BytesReader(buffer);
+
+            var messages = new List<Message>();
+            while (parser.TryParseMessage(ref reader, MessageFormat.Text, out var message))
+            {
+                messages.Add(message);
+            }
+
+            Assert.Equal(reader.Index, Encoding.UTF8.GetByteCount(encoded));
+
+            Assert.Equal(4, messages.Count);
+            MessageTestUtils.AssertMessage(messages[0], MessageType.Binary, new byte[0]);
+            MessageTestUtils.AssertMessage(messages[1], MessageType.Text, "Hello,\r\nWorld!");
+            MessageTestUtils.AssertMessage(messages[2], MessageType.Close, "A");
+            MessageTestUtils.AssertMessage(messages[3], MessageType.Error, "Server Error");
+        }
+
+        [Theory]
+        [InlineData("")]
+        [InlineData("ABC")]
+        [InlineData("1230450945")]
+        [InlineData("5:T:A")]
+        [InlineData("5:T:ABCDE")]
+        public void ReadIncompleteMessages(string encoded)
+        {
+            var parser = new MessageParser();
+            var buffer = Encoding.UTF8.GetBytes(encoded);
+            var reader = new BytesReader(buffer);
+            Assert.False(parser.TryParseMessage(ref reader, MessageFormat.Text, out _));
+        }
+
+        [Theory]
+        [InlineData("5:X:ABCDEF", "Unknown message type: 'X'")]
```

Comment:
The first two are actually Incomplete messages rather than invalid, so they belong in the other test. I'll add them there.